method to generate an initial guess. Fixes #354

Peter230655 opened the pull request:
Is it o.k. to be a method, or should I make it a property?

moorepants commented on opty/direct_collocation.py:
```diff
@@ -686,6 +685,202 @@ def parse_free(self, free):
 
         return parse_free(free, n, q, N, variable_duration)
 
+    def create_initial_guess(self):
```
I think we should call this `create_linear_initial_guess()` or something that is more descriptive. In the future, we may have `create_random_initial_guess` or `create_zero_initial_guess`, etc.

Peter230655 replied:
> I think we should call this `create_linear_initial_guess()` or something that is more descriptive. In the future, we may have `create_random_initial_guess` or `create_zero_initial_guess`, etc.

I will change the name.

moorepants commented on opty/direct_collocation.py:
```diff
@@ -686,6 +685,202 @@ def parse_free(self, free):
         return parse_free(free, n, q, N, variable_duration)
 
+    def create_initial_guess(self):
+        """Creates an initial guess for the free variables in the optimization.
```
```suggestion
        """Returns an initial guess that is the linear interpolation between exact instance constraints otherwise zero."""
```

Peter230655 replied:
This would not be totally correct: if, say, there are bounds ``{u: (a, b)}`` for an unknown trajectory u, my initial guess is ``(a + b)/2``, unless, say, ``b = np.inf``, then the initial guess would be ``a``

moorepants replied:
The first line should say exactly what it returns (as succinctly as possible).

Peter230655 replied:
> The first line should say exactly what it returns (as succinctly as possible).

It is described right below in Notes (before I called it Strategy).

moorepants replied:
But the first line of the docstring should be a succinct description of what it does. Yes it creates an initial guess, but what kind of initial guess does it create?

Peter230655 replied:
> > But the first line of the docstring should be a succinct description of what it does. Yes it creates an initial guess, but what kind of initial guess does it create?
> 
> I changed it like this: `Creates an initial guess that is the linear interpolation between exact instance constraints. Please see the notes for more information.`
Better?

moorepants commented on opty/direct_collocation.py:
```diff
@@ -686,6 +685,202 @@ def parse_free(self, free):
+    def create_initial_guess(self):
+        """Creates an initial guess for the free variables in the optimization.
+
+        Returns:
```
```suggestion
        Returns
```

moorepants commented on opty/direct_collocation.py:
```diff
@@ -686,6 +685,202 @@ def parse_free(self, free):
+        initial_guess : (n*N + q*N + r + s)-ndarray
+            The initial guess for the free variables in the optimization problem.
+
+        Strategy:
```
This is a non-standard heading for numpydoc. It could be "Notes".

Peter230655 replied:
> This is a non-standard heading for numpydoc. It could be "Notes".

I will change it.

moorepants commented on opty/direct_collocation.py:
```diff
@@ -686,6 +685,202 @@ def parse_free(self, free):
+        - All else is set to zero.
+
+    """
+        if self.collocator.instance_constraints is None:
```
indentation seems off. Or maybe indentation of the docstring is off.

Peter230655 replied:
> indentation seems off. Or maybe indentation of the docstring is off.

corrected

moorepants commented:
Is there a good choice of one of the existing examples to add this to? You can show how it is used to the user.

Peter230655 commented:
> Is there a good choice of one of the existing examples to add this to? You can show how it is used to the user.

You mean in examples-gallery?
If yes, surely I will find an example.
Would you have to merge this first, or would I include the example I find in this PR #365?

moorepants commented:
You can include the change to the example here.

moorepants commented on opty/direct_collocation.py:
```diff
@@ -686,6 +685,203 @@ def parse_free(self, free):
+
+        Returns
+        -------
+        initial_guess : (n*N + q*N + r + s)-ndarray
```
```suggestion
        initial_guess : ndarray, shape(n*N + q*N + r + s,)
```

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -1907,3 +1907,405 @@ def test_attributes_read_only():
+
+
+    initial_guess = prob.create_linear_initial_guess()
+    np.testing.assert_allclose(initial_guess, expected_guess)
```
Plot of first guess:

![image](https://github.com/user-attachments/assets/0344155a-2145-47ce-a2e1-d6e2fe30e983)

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -1907,3 +1907,405 @@ def test_attributes_read_only():
+        ux: (-1.0, 1.0),
+        y: (-4.0, 4.0),
+        uy: (-1.0, 1.0),
+    }
```
Easier to compare if placed here:

![image](https://github.com/user-attachments/assets/9d1b6b07-63e4-4f49-b591-9ce63d7b20ad)

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -1907,3 +1907,405 @@ def test_attributes_read_only():
+        x.func(t1) - 1.5*b1,
+        y.func(t1) - 4.0,
+        ux.func(t1) - 1.0+b2,
+        uy.func(t1) - 4.5,
```
In this case I would expect uy to go from 0.0 to 4.5 to 0.0., but you set it to 4.5 to 4.5 to 4.5.

moorepants replied:
i.e. if not specified, the guess should be at zero

moorepants replied:
but maybe this more a design question. if there is only one instance constraint, what assumption to we make about any other values of time?

Peter230655 replied:
> but maybe this more a design question. if there is only one instance constraint, what assumption to we make about any other values of time?

My assumption was that if there is only one instance constraint for a state variable it will have this value at all times. I thought this was the most reasonable one.

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -1907,3 +1907,405 @@ def test_attributes_read_only():
+
+    # A2: np.inf, -np.inf in bounds
+    bounds[a1] = (-np.inf, 10.0)
+    bounds[a2] = (-10.0, np.inf)
```
a1 and a2 are parameters, so it wouldn't affect the initial guess, right? i get same guess as first test:

![image](https://github.com/user-attachments/assets/50f64778-7822-43ee-97a5-222d5d439986)

Peter230655 replied:
a1, a2 are unknown parameters. I set their initial value like this: if there is a bound {a1: (s, t)} I set it at (s + t) /2.
If one of s or t is np.inf I set it to the other (finite) value.
I did not consider bounds which have np.inf on both ends, as they do not make sense in my opinion.

Parameters do not show in the plots.

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -1907,3 +1907,405 @@ def test_attributes_read_only():
+
+        x.func(tf) + 1.0,
+        ux.func(tf) + 5.0,
+    )
```
![image](https://github.com/user-attachments/assets/dc469849-ebd2-4313-aac7-a8d218f80fc4)

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -1907,3 +1907,405 @@ def test_attributes_read_only():
+        y: (-4.0, 4.0),
+        uy: (-1.0, 1.0),
+        h: (1.0, 2.0),
+    }
```
![image](https://github.com/user-attachments/assets/3cc69927-572b-4b44-9d12-f026ccf0e664)

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -1907,3 +1907,405 @@ def test_attributes_read_only():
+
+    # B2: np.inf, -np.inf in bounds
+    bounds[a1] = (-np.inf, 10.0)
+    bounds[a2] = (-10.0, np.inf)
```
![image](https://github.com/user-attachments/assets/336ba99f-2c45-4905-ad21-235ea5d9c70d)

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -1907,3 +1907,405 @@ def test_attributes_read_only():
+    initial_guess = prob.create_linear_initial_guess()
+    np.testing.assert_allclose(initial_guess, expected_guess)
+
+    # B3: no bounds
```
![image](https://github.com/user-attachments/assets/d845a81c-bad7-4695-874c-4b3bca45aaf2)

Peter230655 replied:
These 'peak' plots happen like this If there is no bound for h, h is assumed to be zero.
So, all times ti = integer * h = 0.
The plots for u1, u2 should have a dot at (0 | 0).

moorepants replied:
I guess h should never be zero, right? So maybe we should always assume a value for it...but what value?

Peter230655 replied:
> I guess h should never be zero, right? So maybe we should always assume a value for it...but what value?

I have found, that one should **always** bound h to be non-negative, to avoid senseless 'solutions'.
If there is a bound, say, {h: (0, 0.5)} I set h = 0.25.
So, maybe this would not be a bad default value?

moorepants replied:
Without the user giving some idea of what the total duration of the simulation might be, it's hard (or impossible) to pick a starting number. If someone is simulating molecular dynamics, maybe their duration is a tenth of a second. I guess we could default to some very small value of h. Or this function could take an extract kwarg: `expected_duration=10.0` where the user can give us a bit more info.

moorepants replied:
The expected duration could default to 1.0 second (just as a simple number).

Peter230655 replied:
> The expected duration could default to 1.0 second (just as a simple number).

I will do so.

moorepants replied:
So you like the idea of the kwarg to give a clue on duration?

Peter230655 replied:
> So you like the idea of the kwarg to give a clue on duration?

I like it! I did not think of it, as you normally do not like them.
Default would be false?

moorepants replied:
I don't dislike kwargs.

Default would be `1.0`.

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -1907,3 +1907,405 @@ def test_attributes_read_only():
+
+        x.func(tf0) + 1.0,
+        ux.func(tf0) + 5.0,
+    )
```
![image](https://github.com/user-attachments/assets/2106e957-743f-436e-8565-17d86873b685)

moorepants commented:
I added a commit, so pull this branch before you edit more. You can now do:

```
In [1]: from opty.tests.test_direct_collocation import test_linear_initial_guess

In [2]: test_linear_initial_guess(plot=True)

In [3]: import matplotlib.pyplot as plt

In [4]: plt.show()
```

to quickly plot your guesses.

moorepants commented:
All your do is `git pull origin linear-initial-guess`. Then you are fine to continue.

Peter230655 commented:
> All your do is `git pull origin linear-initial-guess`. Then you are fine to continue.

Thanks!
I would do this when I am in the correct local branch, that is in create-linear-guess ?

moorepants commented:
The branch associated with this PR is named `linear-initial-guess`:

![image](https://github.com/user-attachments/assets/ab5e73db-fd17-4699-8000-475d542b850e)

Peter230655 commented:
Just found a MAJOR plunder! Have to go back to the drawing board.

moorepants commented:
What is the status of this PR? Is it ready for a review?

Peter230655 commented:
> What is the status of this PR? Is it ready for a review?

Yes, I think, it is ready. I tried to test all permutations I could think of.

moorepants commented on opty/direct_collocation.py:
```diff
@@ -686,6 +685,238 @@ def parse_free(self, free):
 
         return parse_free(free, n, q, N, variable_duration)
 
+    def create_linear_initial_guess(self, plot=False):
```
I'm not a fan of this plot kwarg. You can very easily do:

```
guess = problem.create_linear_initial_guess()
problem.plot_trajectories(guess)
```

It is best not to mix calculation methods with plotting methods.

Peter230655 replied:
> I'm not a fan of this plot kwarg. You can very easily do:
> 
> ```
> guess = problem.create_linear_initial_guess()
> problem.plot_trajectories(guess)
> ```
> 
> It is best not to mix calculation methods with plotting methods.

I agree, will correct it.

Peter230655 commented:
- removed the plotting capability.
- added optional kwarg ``end_time``. If given, variable_node_time_interval=1.0/(num_nodes-1) unless bounds for it are given.
- expanded the test to include:
  - no bounds, no instance constraints
  - no bounds
  - different end_times given.

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -1923,3 +1923,526 @@ def test_attributes_read_only():
         with raises(AttributeError):
             setattr(test, XX, 5)
+
+def test_linear_initial_guess(plot=False):
```
I added the plot kwarg here to be able to show all the plots for this test but you seem to have deleted all the plot commands, so now my kwarg does nothing.

moorepants replied:
https://github.com/csu-hmc/opty/pull/365/commits/1a17afb66b5025b8244d8b93a463fc95d75dfdd6

Peter230655 replied:
As I recall, you had the line
```
if plot:
   prob.plot_trajectories(solution)
```
only once, at the very end of the test_linear_initial_guess, and I left it there.

moorepants replied:
I added it so we could easily plot all the test cases.

moorepants replied:
This needs to be resolved, as you deleted my plotting code.

Peter230655 commented:
My local branch was terribly out of sync, but I think I managed to restore it - hopefully anyway.
I added the kwarg plot in test_linear_initial_guess, if set to True, every initial guess will be plotted.

Peter230655 commented:
- I added the plotting feature in the test functions. The title of each plots says, which test it belongs to. 
- If there is a conflict between bounds for the variable time interval and the end_time, bounds have precedence. I added this in the docstrings.

moorepants commented:
I don't think the instance constraints is an accident in any way. Instance constraints are equations, just like the equations that make up the constraints generated from the equations of motion. They can be arbitrary equations, so what would you imagine a dictionary key being?

Peter230655 commented:
> I don't think the instance constraints is an accident in any way. Instance constraints are equations, just like the equations that make up the constraints generated from the equations of motion. They can be arbitrary equations, so what would you imagine a dictionary key being?

I thought like: instance_constraint = {x(t0): -5.0} instead of (x8t0) - 5.0)
Not important, I have to learn how to unravel these sympy terms to some extend anyway.

moorepants commented:
You can do such a thing but what if you have `x(t0) - y(t0) + z(tf) - complex_function_of(parameters`. What is the dictionary key?

Peter230655 commented:
It seems, that with initial_guess = np.zeros(prob.num_nodes) I get warings about zero divides. It keeps on running, but looks 'ugly'
With initial_guess = np.ones(prob.num_nodes) * 0.01 no warnings.
_Should I change to this in my method?_

moorepants commented:
This needs to be updated with master. @Peter230655 what is the status on this PR?

Peter230655 commented:
I will update it.

Status: 
I wanted to add a version with would 'superimpose' random fluctuations on the linear initial guesses.
You said, this might be a good idea, but I should wait until you had reviewed this one.

moorepants commented:
If you want this one in, please update and make sure all is in order. Let me know when you want to review it. I suggest adding random fluctuations in a later PR.

Peter230655 commented:
YES, separate PR for random fluctuations. I di not express myself correctly.

Peter230655 commented:
Just merged with new master.
(When I have to manually resolve conflicts, I tend to destroy about 50%. But it seems to have no conflict)
I should be in order, I believe the tests are pretty exhaustive.

moorepants commented:
Did you restore the plotting I had in the tests? I've asked about that above several times.

Peter230655 commented:
Yes, of course! I did it immediately after I inadevertently da erased this and you told me so.

moorepants commented:
Incorrect formatting:
![image](https://github.com/user-attachments/assets/184cbbc2-6c10-4165-8d19-c061b18f7ebd)

Also, plot the initial guess.

moorepants commented:
The initial guess is not plotted in the sliding block example either.

Peter230655 commented:
No idea what the error was. Removed the plotting as this is the only thing I think I changed.

Peter230655 commented:
@moorepants : Never complain about my German words again! :-)   :-)  !
When I added the plottoing I put
   ```
ax = prob.plot_trajectories(initial_guess)
ax[0].set_title('trajectories of initial guess')       or something similar
```
I got some strange error.

I removed the plotting and it worked
I added the plotting again, but insted of ``ax`` I used ``achsen``, the German word for axes.
Now it works!

moorepants commented:
In the docs build the solution is not optimal:

```
b'Algorithm converged to a point of local infeasibility. Problem may be infeasible.'
```

Peter230655 commented:
I will surely find an example which will work. All I wanted to say was that it works for fewer examples than I had hoped for.
